### PR TITLE
Handle `count * size`

### PR DESCRIPTION
**Pull request by andschwa**
This fixes the deep-copy code to properly handle when both `count` and `size` attributes are specified, plus loosely associated cleanups.

### CODE REVIEW

**Comment by andschwa** [comment]
Part of #1839.

File: tools/oeedger8r/Emitter.ml [hunk 1]
```diff
@@ -154,42 +154,41 @@ let attr_value_to_string argstruct = function
...
-      | None ->
-          let dims =
-            List.map (fun i -> "[" ^ string_of_int i ^ "]") d.array_dims
-          in
-          sprintf "sizeof(%s%s)" type_expr (String.concat "" dims) )
+  let get_ptr_or_decl_size (p : ptr_size) =
+    let size = attr_value_to_string argstruct p.ps_size
+    and count = attr_value_to_string argstruct p.ps_count in
+    match (size, count) with
+    | Some s, None -> s
```
Comment (anakrish) [inline tools/oeedger8r/Emitter.ml]: Nice!

File: tools/oeedger8r/Emitter.ml [hunk 2]
```diff
@@ -154,42 +154,41 @@ let attr_value_to_string argstruct = function
...
-      match attr_value_to_string argstruct p.ps_size with
-      | Some s -> sprintf "(%s / sizeof(%s))" s type_expr
-      | None ->
-          let dims = List.map string_of_int d.array_dims in
-          String.concat " * " dims )
+  let get_ptr_or_decl_count (p : ptr_size) =
+    let size = attr_value_to_string argstruct p.ps_size
+    and count = attr_value_to_string argstruct p.ps_count in
+    match (size, count) with
+    | Some s, None -> sprintf "(%s / sizeof(%s))" s type_expr
```
Comment (anakrish) [inline tools/oeedger8r/Emitter.ml]: We would need a strategy when size is not a multiple of sizeof(type)

File: tools/oeedger8r/Emitter.ml [hunk 3]
```diff
@@ -154,42 +154,41 @@ let attr_value_to_string argstruct = function
...
+    | None, Some c -> c
+    (* TODO: Check that this is an even multiple of the size of type. *)
+    | Some s, Some c -> sprintf "((%s * %s) / sizeof(%s))" s c type_expr
+    | None, None ->
+        let dims = List.map string_of_int decl.array_dims in
```
Comment (anakrish) [inline tools/oeedger8r/Emitter.ml]: Isn't this just get_array_dims like you changed above?

File: tools/oeedger8r/Emitter.ml [hunk 4]
```diff
@@ -377,37 +376,34 @@ let warn_signed_size_or_count_types (fd : func_decl) =
...
+      let id = decl.identifier in
+      if List.mem id size_params then
         match ptype with
-        (* TODO: Combine these two patterns. *)
         | PTVal (Long s | LLong s) when s = Signed ->
-            print_signedness_warning name
-        | PTVal (Int i) when get_int_signedness i = Signed ->
-            print_signedness_warning name
+            print_signedness_warning id
+        | PTVal (Int i) when i.ia_signedness = Signed ->
```
Comment (anakrish) [inline tools/oeedger8r/Emitter.ml]: Not part of this PR, but shouldn't we also warn about int32, int16 etc?

File: tools/oeedger8r/Emitter.ml (hunk 2 above)
Comment (andschwa) [reply tools/oeedger8r/Emitter.ml]: Isn't there a TODO for that right here?

File: tools/oeedger8r/Emitter.ml (hunk 3 above)
Comment (andschwa) [reply tools/oeedger8r/Emitter.ml]: Whoops, that's what happens when I have multiple PRs in progress and have to reorder them 😅

File: tools/oeedger8r/Emitter.ml (hunk 2 above)
Comment (andschwa) [reply tools/oeedger8r/Emitter.ml]: Oh my bad I only put it in the other function 😅

File: tools/oeedger8r/Emitter.ml (hunk 3 above)
Comment (andschwa) [reply tools/oeedger8r/Emitter.ml]: Wait nope, this is doing something entirely different than get_array_dims. This is giving us the expression `2 * 4 * 3 * 5` etc., instead of `[2][4][3][5]`.

File: tools/oeedger8r/Emitter.ml (hunk 2 above)
Comment (andschwa) [reply tools/oeedger8r/Emitter.ml]: Actually it was here, moved it up.
